src/span-data.ts

```diff
@@ -83,18 +95,75 @@ export class SpanData implements SpanDataInterface {
     });
   }
 
-  addLabel(key: string, value: {}) {
+  // tslint:disable-next-line:no-any
+  addLabel(key: string, value: any) {
     const k = traceUtil.truncate(key, Constants.TRACE_SERVICE_LABEL_KEY_LIMIT);
     const stringValue = typeof value === 'string' ? value : util.inspect(value);
     const v = traceUtil.truncate(
         stringValue, traceWriter.get().getConfig().maximumLabelValueSize);
-    this.span.setLabel(k, v);
+    this.span.labels[k] = v;
   }
 
   endSpan() {
-    this.span.close();
-    if (this.isRoot) {
-      traceWriter.get().writeSpan(this);
-    }
+    this.span.endTime = (new Date()).toISOString();
+  }
+}
+
+/**
+ * Represents a real root span, which corresponds to an incoming request.
+ */
+export class RootSpanData extends BaseSpanData {
+  readonly type = SpanDataType.ROOT;
+
+  constructor(
+      trace: Trace, spanName: string, parentSpanId: string,
+      skipFrames: number) {
+    super(trace, spanName, parentSpanId, skipFrames);
+    this.span.kind = SpanKind.RPC_SERVER;
+  }
+
+  endSpan() {
+    super.endSpan();
+    traceWriter.get().writeSpan(this.trace);
   }
 }
+
+/**
+ * Represents a real child span, which corresponds to an outgoing RPC.
+ */
+export class ChildSpanData extends BaseSpanData {
+  readonly type = SpanDataType.CHILD;
+
+  constructor(
+      trace: Trace, spanName: string, parentSpanId: string,
+      skipFrames: number) {
+    super(trace, spanName, parentSpanId, skipFrames);
+    this.span.kind = SpanKind.RPC_CLIENT;
+  }
+}
+
+// Helper function to generate static virtual trace spans.
+const createDummySpanData =
```

Comment:
nit: this would be more readable as:
```ts
function createDummySpanData<T extends SpanDataType>(spanType: T): ... {
  // ...
}
```

Also, `Phantom` is probably a better phrase than `Dummy`.

Reply:
Done